ngraph/core/include/ngraph/descriptor/tensor.hpp

```diff
@@ -74,16 +76,24 @@ namespace ngraph
         protected:
             element::Type m_element_type;
 
-            // TODO(amprocte): For now we are maintaining both m_shape and m_partial_shape fields,
-            //    with m_shape possibly being invalid (get_shape will throw an exception if it
-            //    is). This is because get_shape() returns a const reference. I think ideally we
-            //    should refactor so that get_shape returns by value.
-            Shape m_shape;
+            // TODO: remove along with get_shape
+            // Initially there was ngraph::Shape m_shape only available to keep shape information.
+            // Support for dynamic shapes required transition to ngraph::PartialShape.
+            // To smoothly transition to ngraph::PartialShape we introduced m_partial_shape
+            // and kept m_shape in sync with m_partial_shape. Synchronization point was placed
+            // in set_partial_shape which dramatically affected performance of ngraph::Function
+            // validation. Since we have started the transition to ngraph::PartialShape and reduced
+            // ngraph::Shape usage the only user of m_shape was get_shape method with signature:
+            // const Shape& descriptor::Tensor::get_shape() const
+            // It was decided to move m_shape and m_partial_shape synchronization point there and
+            // to keep methods signature backward compatible.
+            mutable std::mutex shape_mutex;
+            mutable std::atomic_bool m_shape_changed;
+            mutable Shape m_shape;
```

Comment:
Why we need `atomic` and `mutex` here?

Reply:
mutex is needed to protect `m_shape`
It can be used to protect 'm_shape_changed' as well, but in this case `get_shape` will always cause mutex lock which is a bit more expensive than reading of atomic_bool value
`m_shape_changed` cannot be just `bool` because it can cause race condition when 2 threads are executing 'get_shape' simultaneously.
Please note: there is no goal to achieve thread safety between set_partial_shape and get_shape. The goal is to allow calling of 'get_shape' methods in different threads.

Reply:
Thanks for explanation.

Reply:
If assume we might change simultaneously `m_shape` then why we return it by reference in `get_shape`?
If we can change simultaneously `m_partial_shape` (because it change `m_shape`) why is not also protected by mutex?

Reply:
I we assume that `m_shape` can be change simultaneously then why we return it by reference in `get_shape`? This looks like potentially dangling reference.
Why we don't protect `m_partial_shape` any other members by mutex too if class might be accessible simultaneously from different threads?

Reply:
`m_partial_shape` was not originally protected, so this Tensor class cannot be considered as 'thread-safe' at all, and this is not our goal to make it thread-safe.
Also we shall keep API (return by reference) for backward compatibility.
And with this implementation, if multiple threads are calling 'get_shape' simultaneously - only one thread will update `m_shape` (and clear flag m_shape_changed) and other will just return reference to updated m_shape. So goal is to allow multiple threads calling 'get_shape' (it is the case for MULTI plugin for example) - and it is achieved
Speaking about changing shape in one thread and getting it from another - this case will cause problems, because 'Tensor' was never fully thread-safe, so it's client's responsibility to not use 'set' and 'get' methods from other threads